{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eb86d7c5-839b-4a64-90f9-bac72fe25f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "@file:DependsOn(\"com.github.samtools:htsjdk:2.19.0\")\n",
    "@file:DependsOn(\"net.maizegenetics:tassel:5.2.60\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe33a812-795c-4783-95b2-72e472d933ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "import htsjdk.samtools.CigarOperator\n",
    "import htsjdk.samtools.SAMRecord\n",
    "import htsjdk.samtools.SamReaderFactory\n",
    "import htsjdk.samtools.ValidationStringency\n",
    "import net.maizegenetics.util.Utils\n",
    "import java.io.File\n",
    "import net.maizegenetics.util.Sizeof"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b02c0679-9080-4214-ba23-407abc61fdf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data class TranscriptBpAlignmentStats(val xOrEqArray: MutableList<Int>, val eqArray:MutableList<Int>, var nCounts:Int = 0, var dCounts:Int = 0, var numAlignments:Int = 0 )\n",
    "\n",
    "enum class ExonOperator {\n",
    "    L, C, T\n",
    "}\n",
    "val exonOperatorMap = mapOf<Char,ExonOperator>('L' to ExonOperator.L,'C' to ExonOperator.C,'T' to ExonOperator.T)\n",
    "data class ExonBoundary(val length: Int, val operator: ExonOperator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "17b0e463-7025-4b46-8dbc-9e8a9f3c7bcf",
   "metadata": {},
   "outputs": [],
   "source": [
    "/**\n",
    " * Function to parse out the exon boundaries from the transcript name.\n",
    " * They take the form: transcript_1234:100L24C-12C-90C10T\n",
    " */\n",
    "fun parseExonBoundaries(transcriptName: String) : List<List<ExonBoundary>> {\n",
    "    val exonStrings = transcriptName.split(\":\")[1].split(\"-\")\n",
    "\n",
    "    return exonStrings.map { convertExonStringToBoundary(it) }\n",
    "\n",
    "}\n",
    "\n",
    "/**\n",
    " * Function to take a single exon and parse out the operator and the length of that operation.\n",
    " * It needs to have a running sub list of the number characters until it sees an operator.\n",
    " * Then it will convert the temp String into an Int and save out the boundary.\n",
    " */\n",
    "fun convertExonStringToBoundary(exonString: String) : List<ExonBoundary> {\n",
    "    var runningCount = StringBuilder()\n",
    "    val exonBoundaries = mutableListOf<ExonBoundary>()\n",
    "    for(element in exonString) {\n",
    "        //Check to see if we hit an operator\n",
    "        if(element in exonOperatorMap.keys) {\n",
    "            //if so, turn the runningCount into an Int and add the boundary to the list.\n",
    "            val countString = runningCount.toString()\n",
    "            check(!countString.isEmpty()) {\"Error parsing exon name.  No counts for operator\"}\n",
    "\n",
    "            exonBoundaries.add(ExonBoundary(countString.toInt(), exonOperatorMap[element]!!))\n",
    "            runningCount.clear()\n",
    "        }\n",
    "        else {\n",
    "            runningCount.append(\"$element\")\n",
    "        }\n",
    "    }\n",
    "\n",
    "    return exonBoundaries\n",
    "}\n",
    "\n",
    "/**\n",
    " * Function from original method to count each cds bp and aggregate across all the sam files\n",
    " */\n",
    "fun getBpSummary(array: MutableList<Int>): List<Int> {\n",
    "    val countList = mutableListOf<Int>(0,0,0)\n",
    "\n",
    "    for(i in array.indices) {\n",
    "        countList[i%3] += array[i]\n",
    "    }\n",
    "    return countList\n",
    "}\n",
    "\n",
    "/**\n",
    " * Function to parse the SAM records and from the CIGAR string determine mapping and eq bps\n",
    " * This will make an array for both the Mapping Bps and the EQ bps.\n",
    " */\n",
    "fun computeSamBpArrays(samFile: File, minAlignmentPercent : Double = 0.9, cdsConservPercentage : MutableMap<String,Map<String,Double>> = mutableMapOf()) : Map<String,TranscriptBpAlignmentStats> {\n",
    "    println(\"Processing: ${samFile.name}\")\n",
    "    val reader = SamReaderFactory.makeDefault()\n",
    "        .validationStringency(ValidationStringency.SILENT)\n",
    "        .open(samFile)\n",
    "\n",
    "    val samIterator = reader.iterator()\n",
    "\n",
    "    val statMap = mutableMapOf<String, TranscriptBpAlignmentStats>()\n",
    "    val numAlignmentCountMap = mutableMapOf<String,Int>()\n",
    "    val cdsConserveMapByTranscript = mutableMapOf<String,Double>()\n",
    "    while(samIterator.hasNext()) {\n",
    "        val currentRecord = samIterator.next()\n",
    "        //Skip over if its unmapped\n",
    "        if(currentRecord.readUnmappedFlag) {\n",
    "            continue\n",
    "        }\n",
    "\n",
    "//Uncomment for su1 test\n",
    "//        if(currentRecord.readName != \"Zm00001eb174590_T001:224L363C-186C-93C-153C-77C-142C-108C-86C-72C-127C-87C-93C-159C-84C-81C-83C-118C-258C288T\") {\n",
    "//            continue\n",
    "//        }\n",
    "\n",
    "        //Increment the Number of alignments for a given Transcript\n",
    "        numAlignmentCountMap[currentRecord.readName] = (numAlignmentCountMap[currentRecord.readName]?:0)+1\n",
    "\n",
    "        //If its not secondary, we can build the alignmentBp arrays\n",
    "        if(!statMap.containsKey(currentRecord.readName) && !currentRecord.isSecondaryOrSupplementary) {\n",
    "            //Check min\n",
    "            val exonBoundaries = parseExonBoundaries(currentRecord.readName)\n",
    "            val cdsBoundaries = computeCDSPositions(exonBoundaries)\n",
    "            val stats = buildTranscriptBpAlignmentStats(currentRecord)\n",
    "\n",
    "//            println(currentRecord.readName)\n",
    "//            println(exonBoundaries)\n",
    "//            println(stats.xOrEqArray.size)\n",
    "            val numMapping = stats.xOrEqArray.slice(cdsBoundaries.first .. cdsBoundaries.second).sum()\n",
    "\n",
    "            val alignmentPercentage = (numMapping.toDouble()/(cdsBoundaries.second - cdsBoundaries.first + 1))\n",
    "            cdsConserveMapByTranscript[currentRecord.readName] = alignmentPercentage\n",
    "            if(cdsBoundaries.second != cdsBoundaries.first &&  alignmentPercentage >= minAlignmentPercent ) {\n",
    "                statMap[currentRecord.readName] = stats\n",
    "            }\n",
    "\n",
    "        }\n",
    "\n",
    "    }\n",
    "\n",
    "    cdsConservPercentage[samFile.name] = cdsConserveMapByTranscript.toMap()\n",
    "\n",
    "    //Add alignment counters to statmaps\n",
    "    for(key in statMap.keys) {\n",
    "        statMap[key]!!.numAlignments = numAlignmentCountMap[key]?:0\n",
    "    }\n",
    "    return statMap\n",
    "}\n",
    "\n",
    "fun computeCDSPositions(exonBoundaries:List<List<ExonBoundary>>) : Pair<Int,Int> {\n",
    "    val sizesOfOperators = exonBoundaries.flatten()\n",
    "        .groupBy { it.operator }\n",
    "        .map { Pair(it.key, it.value.map { currValue -> currValue.length }.sum()) }\n",
    "        .toMap()\n",
    "\n",
    "    val leaderSize = sizesOfOperators[ExonOperator.L]?:0\n",
    "\n",
    "    val cdsSize = sizesOfOperators[ExonOperator.C]?:0\n",
    "\n",
    "    return Pair(leaderSize, leaderSize + cdsSize - 1)\n",
    "}\n",
    "\n",
    "/**\n",
    " * Function that actually creates the xOrEQ array and the eqArray.  It also counts the number of N and D bps\n",
    " */\n",
    "fun buildTranscriptBpAlignmentStats(samRecord: SAMRecord) : TranscriptBpAlignmentStats {\n",
    "    val xOrEqArray = Array<Int>(samRecord.readLength) { 0 }\n",
    "    val eqArray = Array<Int>(samRecord.readLength) { 0 }\n",
    "\n",
    "    var nCounts = 0\n",
    "    var dCounts = 0\n",
    "\n",
    "    var currentBp = 0\n",
    "    val cigarElements = samRecord.cigar.cigarElements\n",
    "\n",
    "    //Loop through the cigarElements\n",
    "    for(element in cigarElements) {\n",
    "        val operator = element.operator\n",
    "        val count = element.length\n",
    "\n",
    "        if(operator== CigarOperator.N) {\n",
    "            nCounts+=count\n",
    "        }\n",
    "        else if(operator==CigarOperator.D) {\n",
    "            dCounts+=count\n",
    "        }\n",
    "        //Check to see if consumes query\n",
    "        else if(operator.consumesReadBases()) {\n",
    "            //If it consumes read bases, we can walk through the length of the CIGAR operator and set the position as a 1\n",
    "            for(index in 0 until count) {\n",
    "                if(operator.isAlignment) {\n",
    "                    xOrEqArray[currentBp] = 1\n",
    "                }\n",
    "\n",
    "                if (operator==CigarOperator.EQ) {\n",
    "                    eqArray[currentBp] = 1\n",
    "                }\n",
    "\n",
    "                currentBp++\n",
    "            }\n",
    "        }\n",
    "\n",
    "    }\n",
    "\n",
    "    //Check to see if it was reversed during alignment.  If so we need to flip our arrays.\n",
    "    return if(samRecord.readNegativeStrandFlag) {\n",
    "        TranscriptBpAlignmentStats(xOrEqArray.reversed().toMutableList(), eqArray.reversed().toMutableList(), nCounts, dCounts,1)\n",
    "    }\n",
    "    else {\n",
    "        TranscriptBpAlignmentStats(xOrEqArray.toMutableList(), eqArray.toMutableList(),nCounts, dCounts,1)\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "/**\n",
    " * Funtion to aggregate multiple SAM file's TranscriptBpAlignmentStats easily.\n",
    " * Basically it will check to make sure the lengths of the two arrays are the right size and then will sum the values together.\n",
    " */\n",
    "fun addToAggregate(aggregateMap:MutableMap<String, TranscriptBpAlignmentStats>, currentMap:Map<String, TranscriptBpAlignmentStats>) {\n",
    "    for((currentTranscript, alignmentStats) in currentMap) {\n",
    "        if(aggregateMap.containsKey(currentTranscript)) {\n",
    "            check(aggregateMap[currentTranscript]!!.xOrEqArray.size == alignmentStats.xOrEqArray.size) {\"XorEQ arrays are different sizes.  ${currentTranscript}\"}\n",
    "            check(aggregateMap[currentTranscript]!!.eqArray.size == alignmentStats.eqArray.size) {\"EQ arrays are different sizes.  ${currentTranscript}\"}\n",
    "\n",
    "            for(i in alignmentStats.xOrEqArray.indices) {\n",
    "                aggregateMap[currentTranscript]!!.xOrEqArray[i] += alignmentStats.xOrEqArray[i]\n",
    "                aggregateMap[currentTranscript]!!.eqArray[i] += alignmentStats.eqArray[i]\n",
    "            }\n",
    "\n",
    "            aggregateMap[currentTranscript]!!.nCounts += alignmentStats.nCounts\n",
    "            aggregateMap[currentTranscript]!!.dCounts += alignmentStats.dCounts\n",
    "            aggregateMap[currentTranscript]!!.numAlignments += alignmentStats.numAlignments\n",
    "        }\n",
    "        else {\n",
    "            aggregateMap[currentTranscript] = alignmentStats\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4f1a3ca4-0ca3-4868-bab4-4b5262aa34ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "fun countBpsInCDSSamDir(samDir: String, outputFile: String, useOriginalMethod:Boolean,useAllSams:Boolean = false, minCDSPercentage : Double = .90, outputCDSPercentageFile: String) {\n",
    "    val aggregateMap = mutableMapOf<String, TranscriptBpAlignmentStats>()\n",
    "    val samFiles = File(samDir).walk()\n",
    "        .filter { it.isFile }\n",
    "        .filter { !it.isHidden }\n",
    "\n",
    "    val numSams = if(useAllSams) samFiles.toList().count() else -1\n",
    "\n",
    "    val cdsCounter = mutableMapOf<String,Map<String,Double>>()\n",
    "    val samCountByTranscript = samFiles\n",
    "        .map { computeSamBpArrays(it,minCDSPercentage, cdsCounter) }\n",
    "        .map {\n",
    "            addToAggregate(aggregateMap, it)\n",
    "            it.keys\n",
    "        }\n",
    "        .flatten()\n",
    "        .groupingBy { it }\n",
    "        .eachCount()\n",
    "\n",
    "\n",
    "\n",
    "    Utils.getBufferedWriter(outputCDSPercentageFile).use { output ->\n",
    "        val samFiles = cdsCounter.keys.sorted()\n",
    "        output.write(\"TranscriptName\\t${samFiles.joinToString(\"\\t\")}\\n\")\n",
    "\n",
    "        for (transcript in samCountByTranscript.keys.sorted()) {\n",
    "            output.write(\"${transcript}\\t${samFiles.map { cdsCounter[it]?.get(transcript)?:0.0 }.joinToString(\"\\t\")}\\n\")\n",
    "        }\n",
    "    }\n",
    "\n",
    "\n",
    "\n",
    "    //Total % bp1+bp2 and bp3 for each transcript\n",
    "    //transcript\\tlenTranscript\\tnumSams\\tbp1CountM\\tbp2CountM\\tbp3CountM\\tbp1CountEQ\\tbp2CountEQ\\tbp3CountEQ\n",
    "\n",
    "    if(useOriginalMethod) {\n",
    "        outputCountsOriginalMethod(outputFile, samCountByTranscript, aggregateMap)\n",
    "    }\n",
    "    else {\n",
    "        outputCountsFullTranscript(outputFile, samCountByTranscript,numSams, aggregateMap)\n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "/**\n",
    " * Original method to count the bps for CDS based alignments.\n",
    " */\n",
    "fun outputCountsOriginalMethod(\n",
    "    outputFile: String,\n",
    "    samCountByTranscript: Map<String, Int>,\n",
    "    aggregateMap: MutableMap<String, TranscriptBpAlignmentStats>\n",
    ") {\n",
    "    Utils.getBufferedWriter(outputFile).use { output ->\n",
    "        output.write(\n",
    "            \"TranscriptName\\tTranscriptLength\\tNumSams\\t\" +\n",
    "                    \"bp1CountM\\tbp2CountM\\tbp3CountM\\t\" +\n",
    "                    \"bp1CountEQ\\tbp2CountEQ\\tbp3CountEQ\\tnCount\\tdCount\\t\" +\n",
    "                    \"bp1PropM\\tbp2PropM\\tbp3PropM\\ttotalPropM\\t\" +\n",
    "                    \"bp1PropEQ\\tbp2PropEQ\\tbp3PropEQ\\ttotalPropEQ\\tavgNPerSAM\\tavgDPerSAM\\tAvgNumAlignments\\n\"\n",
    "        )\n",
    "\n",
    "        for (transcriptName in samCountByTranscript.keys.sorted()) {\n",
    "            val (xOrEqArray, eqArray, nCount, dCount, numAlignments) = aggregateMap[transcriptName]!!\n",
    "            val xOrEqArrayBpSummary = getBpSummary(xOrEqArray)\n",
    "            val eqArrayBpSummary = getBpSummary(eqArray)\n",
    "            output.write(\n",
    "                \"${transcriptName}\\t${xOrEqArray.size}\\t${samCountByTranscript[transcriptName]}\\t\" +\n",
    "                        \"${xOrEqArrayBpSummary[0]}\\t${xOrEqArrayBpSummary[1]}\\t${xOrEqArrayBpSummary[2]}\\t\" +\n",
    "                        \"${eqArrayBpSummary[0]}\\t${eqArrayBpSummary[1]}\\t${eqArrayBpSummary[2]}\\t${nCount}\\t${dCount}\\t\" +\n",
    "                        \"${3 * (xOrEqArrayBpSummary[0].toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size))}\\t\" +\n",
    "                        \"${3 * (xOrEqArrayBpSummary[1].toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size))}\\t\" +\n",
    "                        \"${3 * (xOrEqArrayBpSummary[2].toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size))}\\t\" +\n",
    "                        \"${(xOrEqArrayBpSummary[0] + xOrEqArrayBpSummary[1] + xOrEqArrayBpSummary[2]).toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size)}\\t\" +\n",
    "                        \"${3 * (eqArrayBpSummary[0].toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size))}\\t\" +\n",
    "                        \"${3 * (eqArrayBpSummary[1].toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size))}\\t\" +\n",
    "                        \"${3 * (eqArrayBpSummary[2].toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size))}\\t\" +\n",
    "                        \"${(eqArrayBpSummary[0] + eqArrayBpSummary[1] + eqArrayBpSummary[2]).toDouble() / (samCountByTranscript[transcriptName]!! * xOrEqArray.size)}\\t\" +\n",
    "                        \"${nCount.toDouble() / samCountByTranscript[transcriptName]!!}\\t\" +\n",
    "                        \"${dCount.toDouble() / samCountByTranscript[transcriptName]!!}\\t\" +\n",
    "                        \"${numAlignments.toDouble() / samCountByTranscript[transcriptName]!!}\" +\n",
    "                        \"\\n\"\n",
    "            )\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "/**\n",
    " * Function to count all the different base pair types for each exon.\n",
    " */\n",
    "fun outputCountsFullTranscript(\n",
    "    outputFile: String,\n",
    "    samCountByTranscript: Map<String, Int>,\n",
    "    samCountByTranscriptDenom: Int,\n",
    "    aggregateMap: MutableMap<String, TranscriptBpAlignmentStats>\n",
    ") {\n",
    "    Utils.getBufferedWriter(outputFile).use { output ->\n",
    "\n",
    "        output.write(\"TranscriptName\\tExonNumber\\tExonLength\\tNumSams\\t\" +\n",
    "                \"num_bp_L\\tnum_bp_C\\tnum_bp_T\\t\" +\n",
    "                \"num_bp_L_Map\\tnum_bp_L_EQ\\t\" +\n",
    "                \"prop_bp_L_Map\\tprop_bp_L_EQ\\t\" +\n",
    "                \"num_bp_T_Map\\tnum_bp_T_EQ\\t\" +\n",
    "                \"prop_bp_T_Map\\tprop_bp_T_EQ\\t\" +\n",
    "                \"num_bp_C_codonPos1_Map\\tnum_bp_C_codonPos1_EQ\\t\" +\n",
    "                \"num_bp_C_codonPos2_Map\\tnum_bp_C_codonPos2_EQ\\t\" +\n",
    "                \"num_bp_C_codonPos3_Map\\tnum_bp_C_codonPos3_EQ\\t\" +\n",
    "                \"prop_bp_C_codonPos1_Map\\tprop_bp_C_codonPos1_EQ\\t\" +\n",
    "                \"prop_bp_C_codonPos2_Map\\tprop_bp_C_codonPos2_EQ\\t\" +\n",
    "                \"prop_bp_C_codonPos3_Map\\tprop_bp_C_codonPos3_EQ\\t\" +\n",
    "                \"num_total_Map\\tnum_total_EQ\\t\" +\n",
    "                \"prop_total_Map\\tprop_total_EQ\\t\" +\n",
    "                \"avgNPerSAM\\tavgDPerSAM\\tAvgNumAlignments\\n\")\n",
    "\n",
    "\n",
    "        for (transcriptName in samCountByTranscript.keys.sorted()) {\n",
    "\n",
    "            val (xOrEqArray, eqArray, nCount, dCount, numAlignments) = aggregateMap[transcriptName]!!\n",
    "\n",
    "\n",
    "\n",
    "            val samCount = samCountByTranscript[transcriptName]!!\n",
    "            val samCountDenom = if(samCountByTranscriptDenom != -1) samCountByTranscriptDenom else samCount\n",
    "            //Parse out the exon numbers:\n",
    "            val exonNumbers = parseExonBoundaries(transcriptName)\n",
    "\n",
    "            var cdsCounter = 0\n",
    "            var currentAlignmentBp = 0\n",
    "\n",
    "            //Create total counters over the full transcript\n",
    "            var totalNumML = 0\n",
    "            var totalNumEqL = 0\n",
    "            var totalNumMT = 0\n",
    "            var totalNumEqT = 0\n",
    "\n",
    "            var totalNumMCBp1 = 0\n",
    "            var totalNumEqCBp1 = 0\n",
    "            var totalNumMCBp2 = 0\n",
    "            var totalNumEqCBp2 = 0\n",
    "            var totalNumMCBp3 = 0\n",
    "            var totalNumEqCBp3 = 0\n",
    "\n",
    "            //Create some counters which are sums of all the different types of poitions we can see.\n",
    "            var totalTranscriptLength = 0\n",
    "            var totalNumL = 0\n",
    "            var totalNumC = 0\n",
    "            var totalNumCBp1 = 0\n",
    "            var totalNumCBp2 = 0\n",
    "            var totalNumCBp3 = 0\n",
    "            var totalNumT = 0\n",
    "\n",
    "            for((index,exon) in exonNumbers.withIndex()) {\n",
    "                //Create temporary  counts for each different type of position within a given exon\n",
    "                var numML = 0\n",
    "                var numEqL = 0\n",
    "                var numMT = 0\n",
    "                var numEqT = 0\n",
    "\n",
    "                var numMCBp1 = 0\n",
    "                var numEqCBp1 = 0\n",
    "                var numMCBp2 = 0\n",
    "                var numEqCBp2 = 0\n",
    "                var numMCBp3 = 0\n",
    "                var numEqCBp3 = 0\n",
    "\n",
    "                var totalLength = 0\n",
    "                var numL = 0\n",
    "                var numC = 0\n",
    "                var numCBp1 = 0\n",
    "                var numCBp2 = 0\n",
    "                var numCBp3 = 0\n",
    "                var numT = 0\n",
    "\n",
    "                //Loop through each operator in the given exon string.\n",
    "                //This should be 1,2 or 3 operators.\n",
    "                for(exonBoundary in exon) {\n",
    "                    val sizeOfExonOp = exonBoundary.length\n",
    "                    totalLength+=sizeOfExonOp //Increment the total length of the exon by the size of this given exon part\n",
    "                    when (exonBoundary.operator) {\n",
    "                        ExonOperator.L -> {\n",
    "                            //Increment the corresponding Leader sequence counts\n",
    "                            numML += xOrEqArray.slice(currentAlignmentBp until currentAlignmentBp+sizeOfExonOp).sum()\n",
    "                            numEqL += eqArray.slice(currentAlignmentBp until currentAlignmentBp+sizeOfExonOp).sum()\n",
    "                            numL += sizeOfExonOp\n",
    "                        }\n",
    "                        ExonOperator.T -> {\n",
    "                            //Increment the corresponding Terminator Sequence counts\n",
    "                            numMT += xOrEqArray.slice(currentAlignmentBp until currentAlignmentBp+sizeOfExonOp).sum()\n",
    "                            numEqT += eqArray.slice(currentAlignmentBp until currentAlignmentBp+sizeOfExonOp).sum()\n",
    "                            numT += sizeOfExonOp\n",
    "                        }\n",
    "                        ExonOperator.C -> {\n",
    "                            //For coding sequence we need to have a running position so we can count bp1, bp2 and bp3 correctly.\n",
    "                            numC += sizeOfExonOp\n",
    "                            for(idx in currentAlignmentBp until currentAlignmentBp+sizeOfExonOp) {\n",
    "                                if(cdsCounter % 3 == 0) {\n",
    "                                    //bp1\n",
    "                                    numMCBp1 += xOrEqArray[idx]\n",
    "                                    numEqCBp1 += eqArray[idx]\n",
    "                                    numCBp1 ++\n",
    "                                }\n",
    "                                else if (cdsCounter % 3 == 1) {\n",
    "                                    //bp2\n",
    "                                    numMCBp2 += xOrEqArray[idx]\n",
    "                                    numEqCBp2 += eqArray[idx]\n",
    "                                    numCBp2++\n",
    "                                }\n",
    "                                else if(cdsCounter % 3 ==2) {\n",
    "                                    //Bp3\n",
    "                                    numMCBp3 += xOrEqArray[idx]\n",
    "                                    numEqCBp3 += eqArray[idx]\n",
    "                                    numCBp3++\n",
    "                                }\n",
    "\n",
    "                                cdsCounter++\n",
    "                            }\n",
    "                        }\n",
    "                    }\n",
    "                    //Shift up the current alignment Base pair by the size of the Exon part.\n",
    "                    currentAlignmentBp += sizeOfExonOp\n",
    "                }\n",
    "\n",
    "                // Increase the total transcript counters correctly\n",
    "                totalNumML += numML\n",
    "                totalNumEqL += numEqL\n",
    "                totalNumMT += numMT\n",
    "                totalNumEqT += numEqT\n",
    "\n",
    "                totalNumMCBp1 += numMCBp1\n",
    "                totalNumEqCBp1 += numEqCBp1\n",
    "                totalNumMCBp2 += numMCBp2\n",
    "                totalNumEqCBp2 += numEqCBp2\n",
    "                totalNumMCBp3 += numMCBp3\n",
    "                totalNumEqCBp3 += numEqCBp3\n",
    "\n",
    "                totalTranscriptLength +=totalLength\n",
    "                totalNumL += numL\n",
    "                totalNumC += numC\n",
    "                totalNumCBp1 += numCBp1\n",
    "                totalNumCBp2 += numCBp2\n",
    "                totalNumCBp3 += numCBp3\n",
    "\n",
    "                totalNumT += numT\n",
    "\n",
    "                //Create total Map and EQ counts for this specific exon\n",
    "                val thisExonTotalMap = numML + numMCBp1 + numMCBp2 + numMCBp3 + numMT\n",
    "                val thisExonTotalEq = numEqL + numEqCBp1 + numEqCBp2 + numEqCBp3 + numEqT\n",
    "\n",
    "\n",
    "\n",
    "                //Write out the output.  Each prop column is scaled by the number of sam files which had an alignment for that transcript\n",
    "                //For the Coding sequence we need to multiply by 3 otherwise it will be 1/3 of what it really is.\n",
    "                //For the average N, D and numAlignments, those values are shared across all exons as they are transcript level measurements.\n",
    "                output.write(\"${transcriptName}\\t${index}\\t${totalLength}\\t${samCount}\\t\" +\n",
    "                        \"${numL}\\t${numC}\\t${numT}\\t\" +\n",
    "                        \"${numML}\\t${numEqL}\\t\" +\n",
    "                        \"${ if(numL== 0) 0.0 else numML.toDouble()/(numL * samCountDenom)}\\t${ if(numL== 0) 0.0 else numEqL.toDouble()/(numL * samCountDenom)}\\t\" +\n",
    "                        \"${numMT}\\t${numEqT}\\t\" +\n",
    "                        \"${ if(numT == 0) 0.0 else numMT.toDouble()/(numT * samCountDenom)}\\t${ if(numT == 0) 0.0 else numEqT.toDouble()/(numT * samCountDenom)}\\t\" +\n",
    "                        \"${numMCBp1}\\t${numEqCBp1}\\t\" +\n",
    "                        \"${numMCBp2}\\t${numEqCBp2}\\t\" +\n",
    "                        \"${numMCBp3}\\t${numEqCBp3}\\t\" +\n",
    "                        \"${ if(numCBp1 == 0) 0.0 else (numMCBp1.toDouble()) / (numCBp1 * samCountDenom)}\\t${ if(numCBp1 == 0) 0.0 else (numEqCBp1.toDouble()) / (numCBp1 * samCountDenom)}\\t\" +\n",
    "                        \"${ if(numCBp2 == 0) 0.0 else (numMCBp2.toDouble()) / (numCBp2 * samCountDenom)}\\t${ if(numCBp2 == 0) 0.0 else (numEqCBp2.toDouble()) / (numCBp2 * samCountDenom)}\\t\" +\n",
    "                        \"${ if(numCBp3 == 0) 0.0 else (numMCBp3.toDouble()) / (numCBp3 * samCountDenom)}\\t${ if(numCBp3 == 0) 0.0 else (numEqCBp3.toDouble()) / (numCBp3 * samCountDenom)}\\t\" +\n",
    "                        \"${thisExonTotalMap}\\t${thisExonTotalEq}\\t\" +\n",
    "                        \"${thisExonTotalMap.toDouble()/(totalLength * samCountDenom)}\\t${thisExonTotalEq.toDouble()/(totalLength* samCountDenom)}\\t\" +\n",
    "                        \"${nCount.toDouble() / samCountDenom}\\t\" +\n",
    "                        \"${dCount.toDouble() / samCountDenom}\\t\" +\n",
    "                        \"${numAlignments.toDouble() / samCountDenom}\" +\n",
    "                        \"\\n\")\n",
    "\n",
    "            }\n",
    "\n",
    "\n",
    "            //Sum up the total Mapping counts and the EQ counts and export the full transcript counts.\n",
    "            val fullTranscriptTotalMap = totalNumML + totalNumMCBp1 + totalNumMCBp2 + totalNumMCBp3 + totalNumMT\n",
    "            val fullTranscriptTotalEq = totalNumEqL + totalNumEqCBp1 + totalNumEqCBp2 + totalNumEqCBp3 + totalNumEqT\n",
    "\n",
    "\n",
    "            output.write(\"${transcriptName}\\tfull\\t${totalTranscriptLength}\\t${samCount}\\t\" +\n",
    "                    \"${totalNumL}\\t${totalNumC}\\t${totalNumT}\\t\" +\n",
    "                    \"${totalNumML}\\t${totalNumEqL}\\t\" +\n",
    "                    \"${ if(totalNumL== 0) 0.0 else totalNumML.toDouble()/(totalNumL * samCountDenom)}\\t${ if(totalNumL== 0) 0.0 else totalNumEqL.toDouble()/(totalNumL * samCountDenom)}\\t\" +\n",
    "                    \"${totalNumMT}\\t${totalNumEqT}\\t\" +\n",
    "                    \"${ if(totalNumT== 0) 0.0 else totalNumMT.toDouble()/(totalNumT * samCountDenom)}\\t${ if(totalNumT== 0) 0.0 else totalNumEqT.toDouble()/(totalNumT * samCountDenom)}\\t\" +\n",
    "                    \"${totalNumMCBp1}\\t${totalNumEqCBp1}\\t\" +\n",
    "                    \"${totalNumMCBp2}\\t${totalNumEqCBp2}\\t\" +\n",
    "                    \"${totalNumMCBp3}\\t${totalNumEqCBp3}\\t\" +\n",
    "                    \"${ if(totalNumCBp1 == 0) 0.0 else (totalNumMCBp1.toDouble()) / (totalNumCBp1 * samCountDenom)}\\t${ if(totalNumCBp1 == 0) 0.0 else (totalNumEqCBp1.toDouble()) / (totalNumCBp1 * samCountDenom)}\\t\" +\n",
    "                    \"${ if(totalNumCBp2 == 0) 0.0 else (totalNumMCBp2.toDouble()) / (totalNumCBp2 * samCountDenom)}\\t${ if(totalNumCBp2 == 0) 0.0 else (totalNumEqCBp2.toDouble()) / (totalNumCBp2 * samCountDenom)}\\t\" +\n",
    "                    \"${ if(totalNumCBp3 == 0) 0.0 else (totalNumMCBp3.toDouble()) / (totalNumCBp3 * samCountDenom)}\\t${ if(totalNumCBp3 == 0) 0.0 else (totalNumEqCBp3.toDouble()) / (totalNumCBp3 * samCountDenom)}\\t\" +\n",
    "                    \"${fullTranscriptTotalMap}\\t${fullTranscriptTotalEq}\\t\" +\n",
    "                    \"${fullTranscriptTotalMap.toDouble()/(totalTranscriptLength * samCountDenom)}\\t${fullTranscriptTotalEq.toDouble()/(totalTranscriptLength * samCountDenom)}\\t\" +\n",
    "                    \"${nCount.toDouble() / samCountDenom}\\t\" +\n",
    "                    \"${dCount.toDouble() / samCountDenom}\\t\" +\n",
    "                    \"${numAlignments.toDouble() / samCountDenom}\" +\n",
    "                    \"\\n\")\n",
    "\n",
    "        }\n",
    "    }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa10da07-6fd0-429e-b9ab-6f945752f69e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------\n",
      "Current Heap Size: 48 MB\n",
      "Max Available Heap: 1946 MB\n",
      "-------------------------------\n",
      "Processing: Bl-K1279B-Draft-PanAnd-1.0.fasta_annotated_cds_eqx.sam\n",
      "Processing: Ud-A1076-Draft-PanAnd-1.0.fasta_annotated_cds_eqx.sam\n",
      "Processing: AN21TS70.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: Vc-Pasquet1098-Draft-PanAnd-1.0.fasta_annotated_cds_eqx.sam\n",
      "Processing: AN21TSTL0199.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN20T021.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN20T022.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN21TS56.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN21TS12.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: Zv-TIL11-Reference-PanAnd-2.0.fasta_annotated_cds_eqx.sam\n",
      "Processing: AN21TNTL0112.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: Zx-TIL18-Reference-PanAnd-2.0.fasta_annotated_cds_eqx.sam\n",
      "Processing: AN21TNTL0174.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: ChrysopogonNigritanus_v1.0.fa_annotated_cds_eqx.sam\n",
      "Processing: AN21TS39.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN20T020.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN21TS35.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN20T194.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN20T036.final.contigs_annotated_cds_eqx.sam\n",
      "Processing: AN21TNTL0137.final.contigs_annotated_cds_eqx.sam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Java heap space\n",
      "java.lang.OutOfMemoryError: Java heap space\n",
      "org.jetbrains.kotlinx.jupyter.ReplEvalRuntimeException: Java heap space\n",
      "\tat org.jetbrains.kotlinx.jupyter.repl.impl.InternalEvaluatorImpl.eval(InternalEvaluatorImpl.kt:106)\n",
      "\tat org.jetbrains.kotlinx.jupyter.repl.impl.CellExecutorImpl$execute$1$result$1.invoke(CellExecutorImpl.kt:64)\n",
      "\tat org.jetbrains.kotlinx.jupyter.repl.impl.CellExecutorImpl$execute$1$result$1.invoke(CellExecutorImpl.kt:63)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ReplForJupyterImpl.withHost(repl.kt:598)\n",
      "\tat org.jetbrains.kotlinx.jupyter.repl.impl.CellExecutorImpl.execute(CellExecutorImpl.kt:63)\n",
      "\tat org.jetbrains.kotlinx.jupyter.repl.CellExecutor$DefaultImpls.execute$default(CellExecutor.kt:13)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ReplForJupyterImpl$evalEx$1.invoke(repl.kt:421)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ReplForJupyterImpl$evalEx$1.invoke(repl.kt:410)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ReplForJupyterImpl.withEvalContext(repl.kt:374)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ReplForJupyterImpl.evalEx(repl.kt:410)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ReplForJupyterImpl.eval(repl.kt:458)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ProtocolKt$shellMessagesHandler$res$1.invoke(protocol.kt:290)\n",
      "\tat org.jetbrains.kotlinx.jupyter.ProtocolKt$shellMessagesHandler$res$1.invoke(protocol.kt:289)\n",
      "\tat org.jetbrains.kotlinx.jupyter.JupyterConnection$runExecution$execThread$1.invoke(connection.kt:166)\n",
      "\tat org.jetbrains.kotlinx.jupyter.JupyterConnection$runExecution$execThread$1.invoke(connection.kt:164)\n",
      "\tat kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:30)\n",
      "Caused by: java.lang.OutOfMemoryError: Java heap space\n",
      "\n"
     ]
    }
   ],
   "source": [
    "//Full Run of 100 Assemblies\n",
    "val fullTranscriptSamDir = \"/home/jupyter-ajs692/shared_data/aimee/privateGenes_ExonModels/minimap_alignments\"\n",
    "\n",
    "val outputFullTranscriptFile = \"/home/jupyter-ajs692/shared_data/aimee/privateGenes_ExonModels/CDSOutputFiles/privateGenesTranscriptCountFile_FixedProportions_90PercentCDSConservation.txt\"\n",
    "\n",
    "val outputCDSCounterFile = \"/home/jupyter-ajs692/shared_data/aimee/privateGenes_ExonModels/CDSOutputFiles/privateGenesTranscriptCountFile_FixedProportions_90PercentCDSConservation_CDSConsMatrix.txt\"\n",
    "\n",
    "Sizeof.printMemoryUse()\n",
    "countBpsInCDSSamDir(fullTranscriptSamDir, outputFullTranscriptFile, false, false, .9, outputCDSCounterFile)\n",
    "Sizeof.printMemoryUse()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69a06013-d0ce-4d0d-9f34-5d9166dbd92b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Kotlin",
   "language": "kotlin",
   "name": "kotlin"
  },
  "language_info": {
   "codemirror_mode": "text/x-kotlin",
   "file_extension": ".kt",
   "mimetype": "text/x-kotlin",
   "name": "kotlin",
   "nbconvert_exporter": "",
   "pygments_lexer": "kotlin",
   "version": "1.6.20-dev-6372"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
